pkg/registry/discovery/endpointslice/strategy.go

```diff
@@ -103,41 +103,55 @@ func (endpointSliceStrategy) AllowUnconditionalUpdate() bool {
 	return true
 }
 
-// dropDisabledConditionsOnCreate will drop the terminating condition if the
-// EndpointSliceTerminatingCondition is disabled. Otherwise the field is left untouched.
-func dropDisabledConditionsOnCreate(endpointSlice *discovery.EndpointSlice) {
-	if utilfeature.DefaultFeatureGate.Enabled(features.EndpointSliceTerminatingCondition) {
-		return
+// dropDisabledConditionsOnCreate will drop any fields that are disabled.
+func dropDisabledFieldsOnCreate(endpointSlice *discovery.EndpointSlice) {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.EndpointSliceTerminatingCondition) {
+		for i := range endpointSlice.Endpoints {
+			endpointSlice.Endpoints[i].Conditions.Serving = nil
+			endpointSlice.Endpoints[i].Conditions.Terminating = nil
+		}
 	}
 
-	// Always drop the serving/terminating conditions on create when feature gate is disabled.
-	for i := range endpointSlice.Endpoints {
-		endpointSlice.Endpoints[i].Conditions.Serving = nil
-		endpointSlice.Endpoints[i].Conditions.Terminating = nil
+	if !utilfeature.DefaultFeatureGate.Enabled(features.EndpointSliceNodeName) {
+		for i := range endpointSlice.Endpoints {
```

Comment:
I was wondering this question to myself as I was working on the `EndpointSliceTerminatingCondition` feature and I think it might be worthwhile to bring it up now. In v1.20 where most clusters will have both the `EndpointSliceTerminatingCondition` and `EndpointSliceNodeName` gates off, could it be problematic performance-wise that on CREATE and UPDATE we iterate over the entire endpoints list for every EndpointSlice two times to drop disabled fields? Or is this negligible given each EndpointSlice will typically not have more than 100 endpoints?

Reply:
and even if it's negligible, is it worth optimizing this so we drop disabled fields in one loop?
```
if utilfeature.DefaultFeatureGate.Enabled(features.EndpointSliceTerminatingCondition) && utilfeature.DefaultFeatureGate.Enabled(features.EndpointSliceNodeName) {
    return
}

dropTerminating := !utilfeature.DefaultFeatureGate.Enabled(features.EndpointSliceTerminatingCondition)
dropNodeName := !utilfeature.DefaultFeatureGate.Enabled(features.EndpointSliceNodeName)
for i := range endpointSlice.Endpoints {
    if dropTerminating {
        endpointSlice.Endpoints[i].Conditions.Serving = nil
        endpointSlice.Endpoints[i].Conditions.Terminating = nil
    }

    if dropNodeName {
        endpointSlice.Endpoints[i].NodeName = nil
    }
}
```

Reply:
same comment for Update

Reply:
Good call, updated to use that approach now.